### PR TITLE
Add vectorization to 4 functions

### LINKED ISSUE #310: vectorize sct2e, scencd, and et2utc
This time, it's a floating point array, so I expected numpy to work, but it doesn't:

```python
arr = np.array([3.9138485783327938e+13, 3.9138485783955961e+13,
       3.9138485784583984e+13, 3.9138485785839891e+13,
       3.9138485786467852e+13])
scid = -64
spice.sct2e(scid, arr)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-77-aee45d30f174> in <module>
----> 1 spice.sct2e(scid, arr)

~/miniconda3/envs/py37/lib/python3.7/site-packages/spiceypy/spiceypy.py in with_errcheck(*args, **kwargs)
     75     def with_errcheck(*args, **kwargs):
     76         try:
---> 77             res = f(*args, **kwargs)
     78             checkForSpiceError(f)
     79             return res

~/miniconda3/envs/py37/lib/python3.7/site-packages/spiceypy/spiceypy.py in sct2e(sc, sclkdp)
  11076     """
  11077     sc = ctypes.c_int(sc)
> 11078     sclkdp = ctypes.c_double(sclkdp)
  11079     et = ctypes.c_double()
  11080     libspice.sct2e_c(sc, sclkdp, ctypes.byref(et))

TypeError: only size-1 arrays can be converted to Python scalars
```

### PR DESCRIPTION
**Pull request by michaelaye**
* using previously established pattern
* fixes #310

### CODE REVIEW

**Comment by AndrewAnnex** [comment]
I added a new function called isiterable to the support_types.py file that should be used instead when a iterable collection of strings is used, that should fix the failed tests...

-Andrew Annex

File: spiceypy/spiceypy.py [hunk 1]
```diff
@@ -5060,7 +5060,7 @@ def et2utc(et, formatStr, prec, lenout=_default_len_out):
     http://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/cspice/et2utc_c.html
 
     :param et: Input epoch, given in ephemeris seconds past J2000.
-    :type et: float
+    :type et: Union[float,Iterable[float]]
```
Comment (AndrewAnnex) [inline spiceypy/spiceypy.py]: update return type

File: spiceypy/spiceypy.py [hunk 2]
```diff
@@ -10976,7 +10984,7 @@ def scencd(sc, sclkch, MXPART=None):
     :param sc: NAIF spacecraft identification code.
     :type sc: int
     :param sclkch: Character representation of a spacecraft clock.
-    :type sclkch: str
+    :type sclkch: Union[str,Iterable[str]]
```
Comment (AndrewAnnex) [inline spiceypy/spiceypy.py]: update return type below

File: spiceypy/spiceypy.py [hunk 3]
```diff
@@ -11070,15 +11086,23 @@ def sct2e(sc, sclkdp):
     :param sc: NAIF spacecraft ID code.
     :type sc: int
     :param sclkdp: SCLK, encoded as ticks since spacecraft clock start.
-    :type sclkdp: float
+    :type sclkdp: Union[float,Iterable[float]]
```
Comment (AndrewAnnex) [inline spiceypy/spiceypy.py]: update return type below

File: spiceypy/spiceypy.py [hunk 4]
```diff
@@ -13361,14 +13385,14 @@ def str2et(time):
...
-    representing the number of TDB seconds past the J2000 epoch 
+    Converts a standard Python datetime to a double precision value
+    representing the number of TDB seconds past the J2000 epoch
     corresponding to the input epoch.
 
     https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/req/time.html#The%20J2000%20Epoch
 
     :param dt: A standard Python datetime
-    :type time: datetime
+    :type time: Union[datetime,Iterable[datetime]]
```
Comment (AndrewAnnex) [inline spiceypy/spiceypy.py]: update return type below

File: spiceypy/spiceypy.py [hunk 5]
```diff
@@ -14480,15 +14504,23 @@ def utc2et(utcstr):
 
     http://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/cspice/utc2et_c.html
 
-    :param utcstr: Input time string, UTC. 
-    :type utcstr: str
+    :param utcstr: Input time string, UTC.
+    :type utcstr: Union[str,Iterable[str]]
```
Comment (AndrewAnnex) [inline spiceypy/spiceypy.py]: update return type below

File: spiceypy/tests/test_wrapper.py [hunk 6]
```diff
@@ -5701,6 +5703,8 @@ def test_scencd():
     sclkdp = spice.scencd(-32, sclkch)
     npt.assert_almost_equal(sclkdp, 985327950.0)
     assert sclkch == "2/20538:39:768"
+    mylist = 3 * [sclkch]
+    npt.assert_almost_equal(spice.scencd(-32, mylist), 3*[985327950.0])
```
Comment (AndrewAnnex) [inline spiceypy/tests/test_wrapper.py]: for these and other new unit tests, they should be independent test functions that include vectorized in the function name ie `test_scencd_vectorized`. see rest of test_wrapper.py for examples.

File: spiceypy/spiceypy.py [hunk 7]
```diff
@@ -10985,8 +10993,16 @@ def scencd(sc, sclkch, MXPART=None):
     sc = ctypes.c_int(sc)
     sclkch = stypes.stringToCharP(sclkch)
     sclkdp = ctypes.c_double()
-    libspice.scencd_c(sc, sclkch, ctypes.byref(sclkdp))
-    return sclkdp.value
+    if hasattr(sclkch, "__iter__"):
```
Comment (AndrewAnnex) [inline spiceypy/spiceypy.py]: for this function and for the other 4, change to use new iterable test function to fix failed builds
```python
if stypes.isiterable(sclkch):
```

**Comment by AndrewAnnex** [comment]
@michaelaye do you have time to update this PR? If not could you enable edits from maintainers so I could wrap it up?

**Comment by michaelaye** [comment]
later today, yes. just finishing an MDAP proposal right now.

File: spiceypy/spiceypy.py (hunk 1 above)
Comment (michaelaye) [reply spiceypy/spiceypy.py]: done

File: spiceypy/spiceypy.py (hunk 2 above)
Comment (michaelaye) [reply spiceypy/spiceypy.py]: done

File: spiceypy/spiceypy.py (hunk 3 above)
Comment (michaelaye) [reply spiceypy/spiceypy.py]: done

File: spiceypy/spiceypy.py (hunk 4 above)
Comment (michaelaye) [reply spiceypy/spiceypy.py]: done

File: spiceypy/spiceypy.py (hunk 5 above)
Comment (michaelaye) [reply spiceypy/spiceypy.py]: done

File: spiceypy/spiceypy.py (hunk 7 above)
Comment (michaelaye) [reply spiceypy/spiceypy.py]: what about the other 14 uses of `if hasattr`?

Comment (michaelaye) [reply spiceypy/spiceypy.py]: I can fix all of them if you want, IF you let me remove those un-holy white-spaces at the end of lines and in empty lines as well? Deal? ;)

Comment (michaelaye) [reply spiceypy/spiceypy.py]: btw, i also found functions where you accept both scalar and iterable, but return type in the docstring does not mention it, e.g. srfxpt . Do those docstrings also need fixing?

Comment (michaelaye) [reply spiceypy/spiceypy.py]: another btw: as we want SpiceyPy to become PEP8 compliant in the future, right, we need to work on some stuff anyway. My editor removes them by default, that's why the white space was gone.

Quote from PEP8:

> Avoid trailing whitespace anywhere. Because it's usually invisible, it can be confusing: e.g. a backslash followed by a space and a newline does not count as a line continuation marker. Some editors don't preserve it and many projects (like CPython itself) have pre-commit hooks that reject it.

Comment (michaelaye) [reply spiceypy/spiceypy.py]: i have updated all `if hasattr` to `stypes.isiterable` if they fit the pattern of looping over a potential iterable, otherwise do skalar type return. There were some cases with cells where that pattern didn't match, so i didn't touch those.

**Comment by michaelaye** [comment]
Seems like tests are still failing, ctypes wants a float not an iterable, it seems? How did you make it work for the other iterables?
```
>       et = ctypes.c_double(et)
456E       TypeError: a float is required
```

File: README.rst [hunk 8]
```diff
@@ -32,7 +32,7 @@ SPICE is an essential tool for scientists and engineers alike in the planetary
 science field for Solar System Geometry. Please visit the NAIF website for more details about SPICE.
 
 *IMPORTANT*: I have no current affiliation with NASA, NAIF, or JPL. The
-code is provided "as is", use at your own risk.
+code is provided "as is", use at your own risk. However, the NAIF now distributes python "lessons" that use SpiceyPy as the python to spice interface.
```
Comment (AndrewAnnex) [inline README.rst]: I have already added this to the master branch, but I am not sure why this diff is still being shown by github.

File: README.rst [hunk 9]
```diff
@@ -97,7 +100,7 @@ as newer versions are released.
 
 - OS: OS X, Linux, Windows
 - CPU: 64bit only!
-- Python 2.7, 3.3, 3.4, 3.5, 3.6, 3.7
+- Python 2.7, 3.5, 3.6, 3.7
```
Comment (AndrewAnnex) [inline README.rst]: okay something strange is happening

File: spiceypy/spiceypy.py [hunk 10]
```diff
@@ -14480,15 +14530,23 @@ def utc2et(utcstr):
...
 
-    :param utcstr: Input time string, UTC. 
-    :type utcstr: str
+    :param utcstr: Input time string, UTC.
+    :type utcstr: Union[str,Iterable[str]]
     :return: Output epoch, ephemeris seconds past J2000.
-    :rtype: float
+    :rtype: Union[float,Iterable[float]]
     """
     utcstr = stypes.stringToCharP(utcstr)
```
Comment (AndrewAnnex) [inline spiceypy/spiceypy.py]: delete this to fix the unit test

**Comment by AndrewAnnex** [comment]
okay, so there a few things going on here. In the future please try to seperate these comments into different posts so I can reply to each individually. 

1. I do not follow PEP8, or I mostly do? Either way, this PR must remain focused on getting these 4 functions working and up to spec with the rest of the codebase and should do nothing else. This policy is to keep the review process short and clear.

2. using hasattr is fine if the collection is a 1d list or array of numerical/boolean datatypes. `stypes.isiterable` should only be used for functions in which a list of strings are given at the moment I have not tested it for other use cases. If that function needs to be updated to be used uniformly across spiceypy let me worry about that in a later commit. 

3. et2utc is failing because you are passing a collection of floats into ctypes.c_double, please look at the other vectorized functions that use hasattr to follow the existing pattern in the codebase.

4. Something is very wrong with sct2e now given the unit test errors I am seeing in the travis CI logs

5. utc2et is failing because you didn't delete the first call to stringToCharP

6. There is something really odd with the git history of this branch now, you pulled in a lot of diffs that I committed into master creating a very complicated diff that is hard to parse through. I think you can fix this by squashing your changes and rebasing onto the master branch then force pushing to overwrite this branch on your fork. 

7. All white space diffs have to be reverted before I merge. I like to keep commits simple and although there might be good arguments to use black or pep8 in spiceypy, this PR is not the time for them. 

8. Unit tests I suggested need to be added to test the vectorization of these 4 functions.

File: spiceypy/tests/test_wrapper.py [hunk 11]
```diff
@@ -5704,6 +5810,16 @@ def test_scencd():
     spice.kclear()
 
 
+def test_scencd_vectorized():
+    spice.kclear()
+    spice.furnsh(CoreKernel.testMetaKernel)
```
Comment (AndrewAnnex) [inline spiceypy/tests/test_wrapper.py]: CoreKernels

File: spiceypy/tests/test_wrapper.py [hunk 12]
```diff
@@ -5747,6 +5863,16 @@ def test_sct2e():
     spice.kclear()
 
 
+def test_sct2e_vectorized():
+    spice.kclear()
+    spice.furnsh(CoreKernels.testMetaKernel)
+    spice.furnsh(ExtraKernels.voyagerSclk)
+    et = spice.sct2e(-32, 985327965.0)
+    mylist = 3 * [985327965.0]
+    assert spice.sct2e(mylist) == 3*[et]
```
Comment (AndrewAnnex) [inline spiceypy/tests/test_wrapper.py]: you are forgetting to pass in -32
